File: src/infinicore/ops/sparse_gather/sparse_gather.cc
#include "infinicore/ops/sparse_gather.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(SparseGather);

SparseGather::SparseGather(Tensor output, const SpVec &pattern, const Tensor &input) {
    INFINICORE_ASSERT(pattern);
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(output, pattern->values(), pattern->indices(), input);
    INFINICORE_GRAPH_OP_DISPATCH(output->device().getType(), output, pattern, input);
}

void SparseGather::execute(Tensor output, const SpVec &pattern, const Tensor &input) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(SparseGather, output, pattern, input);
}

Tensor sparse_gather(const SpVec &pattern, const Tensor &input) {
    INFINICORE_ASSERT(pattern);
    auto output = Tensor::empty({pattern->nnz()}, pattern->dtype(), input->device());
    sparse_gather_(output, pattern, input);
    return output;
}

void sparse_gather_(Tensor output, const SpVec &pattern, const Tensor &input) {
    SparseGather::execute(output, pattern, input);
}

} // namespace infinicore::op

File: python/infinicore/ops/sparse_gather.py
from infinicore.lib import _infinicore
from infinicore.spvec import SpVec
from infinicore.tensor import Tensor


def sparse_gather(pattern: SpVec, input: Tensor, *, out=None):
    if not isinstance(pattern, SpVec):
        raise TypeError("sparse_gather expects a COO SpVec pattern")

    if out is None:
        return Tensor(_infinicore.sparse_gather(pattern._underlying, input._underlying))

    _infinicore.sparse_gather_(
        out._underlying,
        pattern._underlying,
        input._underlying,
    )
    return out
